### DISCUSSION TITLE
canvasMouseVisualizer

### DISCUSSION THREAD

**Discussion by dalilamma** [ttlns/brotector#3] [Q&A]
What's happening here?

What is the significance of this -

CanvasRenderingContext2D.arc has been called with

cordinates +-5px at current mouse position
canvas +-1px covers the whole page
canvas passes pointerEvents through

Why does this detect a bot?

Is this original work or where did you find it?

I can't make sense of it.

Thx

**Comment by ttlns** [comment]
Detects when a bot attempts to add a visualization of how the mouse moves (draws mouse path on a canvas)

**Comment by dalilamma** [comment]
I don't really understand. You're making the bot draw something and if it's a bot the coordinates will be wrong?

**Comment by ttlns** [comment]
No, some bots just add mouse-path visualization lol. 
And this feature detects that stuff

**Reply by dalilamma** [reply]
> No, some bots just add mouse-path visualization lol. And this feature detects that stuff

do you have some examples of bots which this can detect? or a bot framework which add the mouse-path visualization?

i don't think i've ever seen this before so any pointers towards the bots which do this would be helpful.

thanks!

**Reply by ttlns** [reply]
https://github.com/kaliiiiiiiiii/brotector/blob/31ce357413df6f483f761cf744866d86b7e3a1e5/tests/test_playwright.py#L98 for example

**Comment by dalilamma** [comment]
Ah ok, clever

I'm looking at the code and I don't understand why you need to proxy the apply call


```
function hookFunc(obj, func, callback){
    proxy = new Proxy(globalThis[obj].prototype[func], {
      apply: (target, thisArg, argumentsList) => {
        callback(target, thisArg, argumentsList)
        return Reflect.apply(target, thisArg, argumentsList)
      }
        }
    );
    Object.defineProperty(globalThis[obj].prototype, func, {value:proxy})
}
```

Why don't you just draw the arc normally and look at the mouse position?

**Reply by ttlns** [reply]
This does not draw an arc, but hooks the function
